### PR TITLE
Multi corpus v2

### PR DESCRIPTION
**Pull request by AlexArtrip**
Implementation of a basic multi-corpus search. 

Also, modified the **recall** dp endpoint to include the scores of the retrieved results.

### CODE REVIEW

File: memas/corpus/corpus_searching.py [hunk 1]
```diff
@@ -1,17 +1,55 @@
...
-from memas.interface.corpus import Corpus, CorpusFactory
+from memas.interface.corpus import Corpus, CorpusFactory, CorpusType
 from memas.interface.corpus import Citation
+from collections import defaultdict 
 from memas.interface.storage_driver import DocumentEntity
 from memas.interface.exceptions import SentenceLengthOverflowException
 
 
-def corpora_search(corpus_ids: list[UUID], clue: str) -> list[tuple[float, str, Citation]]:
+def mult_corpus_search(corpus_sets : dict[Corpus], clue, ctx, result_limit) -> list[tuple[float, str, Citation]]:
```
Comment (maxyu1115) [inline memas/corpus/corpus_searching.py:11]: nit: multi

File: memas/dataplane.py [hunk 2]
```diff
@@ -14,21 +18,31 @@ def recall():
...
+    for corpus_info in corpus_infos :
+        corpus_type = corpus_info.corpus_type
         corpus: Corpus = ctx.corpus_provider.get_corpus_by_info(corpus_info)
-        search_results.extend(corpus.search(clue=clue))
+        corpora_grouped_by_type[corpus_type].append(corpus)
 
-    # Combine the results and only take the top ones
-    search_results.sort(key=lambda x: x[0], reverse=True)
+    # Execute a multicorpus search
+    # Need to refactor to remove ctx later and have a cleaner solution, but thats time i dont have right now : (
```
Comment (maxyu1115) [inline memas/dataplane.py:38]: nit: add TODO

File: memas/corpus/corpus_searching.py (hunk 1 above)
Comment (maxyu1115) [inline memas/corpus/corpus_searching.py:11]: nit: incomplete type, what type are the keys of the corpus_sets dict?

File: memas/corpus/corpus_searching.py [hunk 3]
```diff
@@ -1,17 +1,55 @@
...
+    results = defaultdict(list)
+
+    # Direct each multicorpus search to the right algorithm
+    for corpusType, corpora_list in corpus_sets.items() :
```
Comment (maxyu1115) [inline memas/corpus/corpus_searching.py:15]: nit: rename variable name to comply with python standardss

File: memas/corpus/corpus_searching.py [hunk 4]
```diff
@@ -1,17 +1,55 @@
...
+            combined_results.append(sorted_results_matrix[i][j])
+        if len(combined_results) >= result_limit:
+            break
+
+    return combined_results
+
+"""
+All corpora here should be of the same CorpusType implementation (basic_corpus)
+"""
+def basic_corpora_search(corpora: list[Corpus], clue: str, ctx) -> list[tuple[float, str, Citation]]:
```
Comment (maxyu1115) [inline memas/corpus/corpus_searching.py:45]: this entire function looks pretty sus... It looks like we're jumping out of the corpus implementation and redoing the basic corpus search again? **We can keep this for now**, but it'd be best to implement this in a way that properly modularizes the logic (feel free to refactor any interfaces that get in the way).

File: memas/dataplane.py [hunk 5]
```diff
@@ -14,21 +18,31 @@ def recall():
...
+    search_results = mult_corpus_search(corpora_grouped_by_type, clue, ctx, 4)
+    current_app.logger.debug(f"Search Results are: {search_results}")
 
     # TODO : It will improve Query speed significantly to fetch citations after determining which documents to send to user
 
     # Take only top few scores and remove scoring element before sending
-    return [{"document": doc, "citation": asdict(citation)} for doc, citation in search_results[0:5]]
+    return [{"score" : score, "document": doc, "citation": asdict(citation)} for score, doc, citation in search_results[0:5]]
```
Comment (maxyu1115) [inline memas/dataplane.py:45]: Do we want to expose the scores of the search results? What benefits does it add?

File: memas/corpus/corpus_searching.py [hunk 6]
```diff
@@ -1,17 +1,55 @@
...
+    # To combine results for corpora that don't have compareable scoring take equal sized subsets of each Corpus type
+    # TODO : Consider changing this at some point in the future to have better searching of corpus sets with non-comparable scoring
+    combined_results = []
+    for j in range(max([len(x) for x in sorted_results_matrix])) :
+        for i in range(len(sorted_results_matrix)) :
+            if j >= len(sorted_results_matrix[i]) or len(combined_results) >= result_limit:
+                break
+            combined_results.append(sorted_results_matrix[i][j])
+        if len(combined_results) >= result_limit:
+            break
```
Comment (maxyu1115) [inline memas/corpus/corpus_searching.py:38]: This segment for combining the results confuse me. What's the purpose of these two loops, is it to extract an equal number of results from each corpus type equally? In which case wouldn't just extracting the top `result_limit/n`, where `n` is the number of corpus types?

File: memas/dataplane.py (hunk 5 above)
Comment (AlexArtrip) [reply memas/dataplane.py:45]: I'm also a little iffy on whether exposing scoring to users is worth doing, so I leave it up to you. 

I was thinking maybe we do want to expose the results to give developers more insight into what exactly they are retrieving and how they MAY compare with one another. When the corpora scoring isn't comparable it doesn't have obvious value, but I was thinking of how ElasticSearch queries return with scores as well (that can't necessarily be compared with other ES queries).

File: memas/corpus/corpus_searching.py (hunk 6 above)
Comment (AlexArtrip) [reply memas/corpus/corpus_searching.py:38]: The two loops are for extracting result_limit results, preferably equally for each different (non-comparable) way of corpus scoring.  Your suggestion was my initial plan for how to do it, but it gets more complicated when you can't guarantee there are result_limit/n results to fetch from each corpus. The loops are one way of dealing with that while also ordering the results.

File: memas/corpus/corpus_searching.py (hunk 4 above)
Comment (AlexArtrip) [reply memas/corpus/corpus_searching.py:45]: You're right that it is basically a redo of basic corpus search, which isn't ideal. Of course modular code is the way to go, and I was thinking the same thing when I worked on that function. It isn't straightforward without a larger refactor, which I didn't want to do before talking to you. I can work on a possible refactor for that and then we can discuss it.

Comment (maxyu1115) [reply memas/corpus/corpus_searching.py:45]: Ah gotcha, then sure let's keep this for now

File: memas/dataplane.py (hunk 5 above)
Comment (maxyu1115) [reply memas/dataplane.py:45]: Personally think let's leave it out for now, unless there's a strong demand. Main reason being it complicates the interpretation of these results, and is something we can add quite easily

File: memas/corpus/corpus_searching.py (hunk 6 above)
Comment (maxyu1115) [reply memas/corpus/corpus_searching.py:38]: Gotcha, sounds good

**Review by maxyu1115 — APPROVED** [review]

**Pull request merged**